java/code/src/com/redhat/rhn/manager/audit/CVEAuditManagerOVAL.java

```diff
@@ -0,0 +1,295 @@
+     * given CVE identifier as well as sets of relevant channels and erratas.
+     * */
+    public static CVEAuditSystemBuilder doAuditSystem(String cveIdentifier,
+                                                      List<CVEAuditManager.CVEPatchStatus> results,
+                                                      Server clientServer) {
+        // It's possible to have 2 or more patches for one package. It's necessary to apply all of them because
+        // they will have the same outcome i.e. patch the package; instead we need to choose only one.
+        // To choose the one, we rank patches based on the channel they come from .e.g.
+        // assigned, successor product, etc. And for each vulnerable package we keep only the highest ranking patch
+        results = keepOnlyPatchCandidates(results);
+
+        CVEAuditSystemBuilder cveAuditServerBuilder = new CVEAuditSystemBuilder(clientServer.getId());
+        cveAuditServerBuilder.setSystemName(clientServer.getName());
+
+        List<ShallowSystemPackage> allInstalledPackages =
+                PackageManager.shallowSystemPackageList(clientServer.getId());
+
+        LOG.error("Vul packages before filtering: {}",
+                OVALCachingFactory.getVulnerablePackagesByProductAndCve(clientServer.getCpe(), cveIdentifier));
+
+        Set<VulnerablePackage> clientProductVulnerablePackages =
+                OVALCachingFactory.getVulnerablePackagesByProductAndCve(clientServer.getCpe(), cveIdentifier).stream()
+                        .filter(pkg -> isPackageInstalled(pkg, allInstalledPackages))
+                        .collect(Collectors.toSet());
+
+        LOG.error("Vul packages: {}", clientProductVulnerablePackages);
+
+        if (clientProductVulnerablePackages.isEmpty()) {
+            cveAuditServerBuilder.setPatchStatus(PatchStatus.NOT_AFFECTED);
+            return cveAuditServerBuilder;
+        }
+
+        Set<VulnerablePackage> patchedVulnerablePackages = clientProductVulnerablePackages.stream()
+                .filter(vulnerablePackage -> vulnerablePackage.getFixVersion().isPresent()).collect(
+                        Collectors.toSet());
+
+        Set<VulnerablePackage> unpatchedVulnerablePackages = clientProductVulnerablePackages.stream()
+                .filter(vulnerablePackage -> vulnerablePackage.getFixVersion().isEmpty()).collect(
+                        Collectors.toSet());
+
+        boolean allPackagesUnpatched = unpatchedVulnerablePackages.size() == clientProductVulnerablePackages.size();
+
+        if (allPackagesUnpatched) {
+            cveAuditServerBuilder.setPatchStatus(PatchStatus.AFFECTED_PATCH_UNAVAILABLE);
+        }
+        else {
+            boolean allPackagesPatched = patchedVulnerablePackages.stream().allMatch(patchedPackage ->
+                    allInstalledPackages.stream()
+                        .filter(installedPackage ->
```

Comment:
This filter is similar to the one at line 146. Should we use the same method?

Reply:
Nice catch but I don't think we can use the same method. In line 172, in addition to checking if a certain package is installed we also need to access the list of installed versions to be able to assert that all installed versions of the package are above the patched version, which is not possible with `isPackageInstalled`. I created another method to do exactly that.

Reading through the code at CVEAuditManagerOVAL after a while since I last touched it wasn't straightforward; it is not as clear as I hoped for. 

Note to self: Create smaller methods to abstract away common operations and make the code clearer.